### PR TITLE
[quant] fix int16 quantization scale in conv weight

### PR DESCRIPTION
**Pull request by terrychenism**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74665

Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

### CODE REVIEW

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Comment by vkuzo** [comment]
> fix int16 quantization scale in conv weight

In the PR summary, could we provide some context on what specifically was broken, and how you are fixing it?

File: test/quantization/eager/test_quantize_eager_ptq.py [hunk 1]
```diff
@@ -199,6 +199,61 @@ def test_linear(self):
...
+                x = self.quant2(x)
+                x = self.dequant2(x)
+                return x
+
+
+        input_size = (16, 1, 10, 10)
+        data = torch.randn(*input_size, dtype=torch.float)
+
+        original_ref_m = RefM()
+        torch.quantization.engine = "qnnpack"
```
Comment (vkuzo) [inline test/quantization/eager/test_quantize_eager_ptq.py:228]: note: not all test environments will support qnnpack.  You can check for this with `"qnnpack" in torch.testing._internal.common_quantized.supported_qengines`, and skip the test if this is false

File: torch/nn/quantized/_reference/modules/utils.py [hunk 2]
```diff
@@ -98,11 +98,11 @@ def _quantize_weight(
         return weight
 
     if weight_qscheme == torch.per_tensor_affine:
-        if weight_dtype in [torch.quint8, torch.qint8, torch.qint32]:
+        if weight_dtype in [torch.quint8, torch.qint8, torch.qint32, torch.qint32]:
```
Comment (jerryzh168) [inline torch/nn/quantized/_reference/modules/utils.py:101]: nit: torch.qint32 already included before for this one

File: torch/nn/quantized/_reference/modules/utils.py [hunk 3]
```diff
@@ -16,7 +16,7 @@ def _init_weight_qparams(self, weight_qparams, device):
             None, torch.per_tensor_affine, torch.per_channel_affine,
             torch.per_channel_affine_float_qparams], \
             Exception(f"qscheme: {self.weight_qscheme} is not support in reference quantized {self._get_name()}")
-        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2]:
+        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2, torch.qint32]:
```
Comment (vkuzo) [inline torch/nn/quantized/_reference/modules/utils.py:19]: can we add a warning on L37 in the else branch to enable the user to find this easier next time?

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

File: torch/nn/quantized/_reference/modules/utils.py (hunk 3 above)
Comment (jerryzh168) [reply torch/nn/quantized/_reference/modules/utils.py:23]: I think we can add an assert, the `else` branch should not be called

File: test/quantization/eager/test_quantize_eager_ptq.py [hunk 4]
```diff
@@ -199,6 +199,64 @@ def test_linear(self):
...
+        # quantize the reference model
+        original_ref_m.eval()
+        original_ref_m.qconfig = custom_qconfig
+
+        ref_m = prepare(original_ref_m)
+        # calibration
+        ref_m(torch.randn(*input_size, dtype=torch.float))
+
+        ref_m = convert(ref_m, is_reference=True)
+        self.assertNotEqual(ref_m.conv.weight_scale, 1.0)
```
Comment (jerryzh168) [inline test/quantization/eager/test_quantize_eager_ptq.py:257]: hmmm, I think a better test is to initialize a weight observer here and calculate the scale and check it is the same as ref_m.conv.weight_scale

File: torch/nn/quantized/_reference/modules/utils.py [hunk 5]
```diff
@@ -35,6 +39,7 @@ def _init_weight_qparams(self, weight_qparams, device):
                 self.register_buffer(
                     "weight_axis", torch.tensor(0, dtype=torch.int, device=device))
         else:
+            logger.warning("unexpected supported weight dtype: {}".format(self.weight_dtype))
```
Comment (jerryzh168) [inline torch/nn/quantized/_reference/modules/utils.py:42]: I think we can change this to assert, this code is not expected to be executed

File: torch/nn/quantized/_reference/modules/utils.py [hunk 6]
```diff
@@ -16,7 +20,7 @@ def _init_weight_qparams(self, weight_qparams, device):
             None, torch.per_tensor_affine, torch.per_channel_affine,
             torch.per_channel_affine_float_qparams], \
             Exception(f"qscheme: {self.weight_qscheme} is not support in reference quantized {self._get_name()}")
-        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2]:
+        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2, torch.qint32]:
```
Comment (jerryzh168) [inline torch/nn/quantized/_reference/modules/utils.py:23]: actually I feel we can just change this `if` to an assert and remove the else branch..

**Review by jerryzh168 — APPROVED** [review]
looks good overall, had two inline comments, please address them before landing

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

File: torch/nn/quantized/_reference/modules/utils.py (hunk 3 above)
Comment (terrychenism) [reply torch/nn/quantized/_reference/modules/utils.py:19]: I added 'if-else' back due to one of fx test cases needs torch.float, so it will call 'else'

Comment (terrychenism) [reply torch/nn/quantized/_reference/modules/utils.py:19]: TestQuantizeFx.test_dynamic_with_fusion_multiple_uses

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

**Pull request closed without merging**